extension/inet/inet_escape_functions.cpp

```diff
@@ -0,0 +1,491 @@
+	static bool is_control_point(Match *match, const char *input_data) {
+		auto size = match->size;
+		if (size != 4) {
+			return false;
+		}
+
+		auto start = match->start;
+		auto decode_control_point = [&]() {
+			char *hex_input_data = new char[4];
+			hex_input_data[0] = '\\';
+			hex_input_data[1] = input_data[start + 1];
+			hex_input_data[2] = input_data[start + 2];
+			hex_input_data[3] = input_data[start + 3];
+			match->decoded = string_t(hex_input_data, 4);
+			delete[] hex_input_data;
+		};
+
+		if (input_data[start + 2] == '8') { // 0x81, 0x8d, 0x8D, 0x8f, 0x8F
+			if (input_data[start + 3] == '1' || input_data[start + 3] == 'd' || input_data[start + 3] == 'D' ||
+			    input_data[start + 3] == 'f' || input_data[start + 3] == 'F') {
+				decode_control_point();
+				return true;
+			}
+		}
+		if (input_data[start + 2] == '9') { // 0x90, 0x9d, 0x9D
+			if (input_data[start + 3] == '0' || input_data[start + 3] == 'd' || input_data[start + 3] == 'D') {
+				decode_control_point();
+				return true;
+			}
+		}
+		return false;
+	}
+};
+
+template <class OP>
+class HexMatch : public Match {
+public:
+	HexMatch(const char *input_data, UnescapeBindData &bind_data, idx_t beg, idx_t len, bool semi)
+	    : Match(beg, len, semi) {
+		if (OP::check_and_cast_input(this, input_data, num, num_of_chars_to_decode)) {
+			// true, if input is a special case and the value has been already decoded
+			return;
+		}
+		// no need for decoding
+		if (num == 0 && num_of_chars_to_decode != size - 1) {
+			decoded = string_t(input_data + start - 1, size + semi + 1);
+			return;
+		}
+		decode_match(bind_data, decoded);
+
+		// add the rest of the match (if exists) e.g. &#8b; -> valid_part_to_decode(&#8) + add_rest('b;')
+		if (size - 1 != num_of_chars_to_decode) {
+			include_the_rest(input_data);
+		}
+	}
+
+private:
+	void decode_match(UnescapeBindData &bind_data, string_t &decoded) {
+		// line feed character
+		if (num == 0x0d) {
+			decoded = string_t("\\r");
+			return;
+		}
+		// return the replacement character
+		if (num == 0x00 || (0xD800 <= num && num <= 0xDFFF) || 0x10FFFF < num) {
+			decoded = string_t("\uFFFD");
+			return;
+		}
+		// special character references (control points have been already handled at this point)
+		if (0x80 <= num && num <= 0x9f) {
+			D_ASSERT(bind_data.special_invalid_charrefs.find(num) != bind_data.special_invalid_charrefs.end());
+			auto ch = bind_data.special_invalid_charrefs.at(num);
+			decoded = string_t(ch);
+			return;
+		}
+		// skip invalid codepoints
+		if (bind_data.invalid_codepoints.find(num) != bind_data.invalid_codepoints.end()) {
+			decoded = string_t("", 0);
+			return;
+		}
+		char utf8_val[4] = {'\0', '\0', '\0', '\0'};
+		idx_t sz = to_utf8(num, utf8_val);
+		decoded = string_t(utf8_val, sz);
+	}
+
+	void include_the_rest(const char *input_data) {
+		idx_t new_size = decoded.GetSize() + (size + semicolon - num_of_chars_to_decode - 1);
+		char *final_result = new char[new_size];
```

Comment:
`string_t` does not take ownership of any memory passed in, this is leaking the `final_result` currently